contracts/starknet/space/space.cairo

```diff
@@ -0,0 +1,256 @@
+                # TODO: what should we do about the carry
+                let (power, carry) = uint256_add(_abstain, voting_power)
+                power_abstain.write(proposal_id, power)
+                tempvar range_check_ptr = range_check_ptr
+                tempvar syscall_ptr = syscall_ptr
+                tempvar pedersen_ptr = pedersen_ptr
+                tempvar voting_power = voting_power
+            else:
+                # choice is not a valid choice ?!
+                return ()
+            end
+        end
+    end
+
+    let vote = Vote(choice=choice, voting_power=voting_power)
+    vote_registry.write(proposal_id, voter_address, vote)
+
+    return ()
+end
+
+# TODO: execution_hash should be of type Hash and metadata_uri of type felt* (string)
+@external
+func propose{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        proposer_address : EthAddress, execution_hash : felt, metadata_uri : felt) -> ():
+    alloc_locals
+
+    # Verify that the caller is the authenticator contract.
+    authenticator_only()
+
+    let (current_block) = get_block_number()
+    let (delay) = voting_delay.read()
+    let (duration) = voting_period.read()
+
+    # Define start_block and end_block based on current block, delay and duration variables.
+    let start_block = current_block + delay
+    let end_block = start_block + duration
+
+    # Get the voting power of the proposer
+    let (strategy_contract) = voting_strategy.read()
+    let (local params : felt*) = alloc()
+    let (voting_power) = IVotingStrategy.get_voting_power(
+        contract_address=strategy_contract,
+        address=proposer_address,
+        at=current_block,
+        params_len=0,
+        params=params)
+
+    # Verify that the proposer has enough voting power to trigger a proposal
+    let (threshold) = proposal_threshold.read()
+    let (lower) = uint256_lt(voting_power, threshold)
+    if lower == 1:
+        # Not enough voting power to create a proposal
+        assert 1 = 0
+    end
+
+    # Create the proposal and its proposal id
+    let proposal = Proposal(execution_hash, metadata_uri, start_block, end_block)
+    let (proposal_id) = next_proposal_nonce.read()
+
+    # Store the proposal
+    proposal_registry.write(proposal_id, proposal)
+
+    # Increase the proposal nonce
+    next_proposal_nonce.write(proposal_id + 1)
+
+    return ()
+end
+
+@view
+func get_vote_info{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        voter_address : EthAddress, proposal_id : felt) -> (vote : Vote):
+    return vote_registry.read(proposal_id, voter_address)
+end
+
+@view
+func get_nonce{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}() -> (
+        next_nonce : felt):
+    let (nonce) = next_proposal_nonce.read()
+    return (next_nonce=nonce)
+end
+
+@view
+func get_delay{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}() -> (
```

Comment:
If it's just for UI we may not need these views functions, we would just need to emit events with the space settings and the API would store this information.

Reply:
Yeah I initially wrote those with testing in mind. I've removed everything except `get_vote_info` and `get_proposal_info` in [2a1fde7](https://github.com/snapshot-labs/sx-core/pull/47/commits/2a1fde787effff41c2b618dd0c058d31e57020c4) :)